[rerun-sdk] New port at version 0.22.1

adolfos94 opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

adolfos94 committed:
[rerun] Add version 0.22.1

dg0yt commented on ports/rerun/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
+vcpkg_download_distfile(
+    ARCHIVE
+    URLS "https://github.com/rerun-io/rerun/releases/download/${VERSION}/rerun_cpp_sdk.zip" # Replace with the actual URL of the zip file
+    FILENAME rerun_cpp_sdk.zip
+    SHA512 1351dd0937d6ddf73622b69a803a7233eb92e5ec52607fc1c775accd015d52eaf3259c0aea64cfac3109f1c55218fb6a4597bff5b067ccdd194cd8695b3f4c8c
+)
+
+# Extract the zip file
+vcpkg_extract_source_archive(
+    SOURCE_PATH
+    ARCHIVE "${ARCHIVE}"
+)
```
This could probaly be ported to `vcpkg_from_github`.

adolfos94 replied:
It does weird things if I try to get it from github.

dg0yt replied:
Yes. Vendored deps, and rust/cargo.

dg0yt commented on ports/rerun/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS -DRERUN_DOWNLOAD_AND_BUILD_ARROW=OFF
```
Usually formatted like:
```suggestion
    OPTIONS
        -DRERUN_DOWNLOAD_AND_BUILD_ARROW=OFF
```

dg0yt commented on ports/rerun/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+    OPTIONS -DRERUN_DOWNLOAD_AND_BUILD_ARROW=OFF
+)
+
+message(STATUS "Packages directory: ${PACKAGE_PREFIX_DIR}")
+
```
```suggestion
```
There is no such variable in script mode (`portfile.cmake`).
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#useful-implementation-notes

dg0yt commented on ports/rerun/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+
+vcpkg_cmake_install()
+
+vcpkg_cmake_config_fixup(PACKAGE_NAME rerun CONFIG_PATH "lib/cmake/rerun_sdk")
```
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME rerun_sdk CONFIG_PATH "lib/cmake/rerun_sdk")
```
According to the following code, the main file is `rerun_sdkConfig.cmake`, so it must be in a `rerun_sdk[...]` directory.

dg0yt commented on ports/rerun/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+
+vcpkg_replace_string(
+    "${CURRENT_PACKAGES_DIR}/share/rerun/rerun_sdkConfig.cmake"
+    "${SOURCE_PATH}/lib/librerun_c__macos_arm64.a"
```
Which pattern do you try to match here?
`${SOURCE_PATH}` is not where artifacts are built or staged, so I would be surprise to see this appear in the installed file.
OTOH the config templates does several ... unusual ... things.
https://github.com/rerun-io/rerun/blob/main/rerun_cpp/Config.cmake.in

adolfos94 replied:
For some reason, it inserts an absolute path pointing to `librerun_c`.  I'm just replacing the absolute path with a relative path.

adolfos94 committed:
Update vcpkg.json

adolfos94 committed:
Add support static

adolfos94 committed:
Merge branch 'microsoft:master' into rerun

adolfos94 committed:
Add supported architectures

adolfos94 committed:
Update vcpkg.json

adolfos94 committed:
Update rerun.json

adolfos94 commented:
I think I’m going to rename the port from ‘rerun’ to ‘rerun-sdk’

dg0yt commented:
> I think I’m going to rename the port from ‘rerun’ to ‘rerun-sdk’

Good point. On repology.org, `rerun` is already taken by another package, but there is `(python:)rerun-sdk`.
https://repology.org/projects/?search=rerun

adolfos94 committed:
Renamed port to rerun-sdk

adolfos94 committed:
Update portfile.cmake

adolfos94 committed:
Update rerun-sdk.json

adolfos94 committed:
Fix x64-windows build

adolfos94 committed:
Update rerun-sdk.json

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -1,3 +1,7 @@
+# Enforce static linking and compilation
+set(VCPKG_CRT_LINKAGE static)
+set(VCPKG_LIBRARY_LINKAGE static)
```
No, this is not how it works. These variables are controlled by the triplet. Some variations are allowed with `vcpkg_check_linkage`.

There might be a CRT linkage problem due to the prebuilt C SDK library, at least if it needs CRT functions. (I is rust, so I don't know.)

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -1,3 +1,7 @@
+set(VCPKG_CRT_LINKAGE static)
+set(VCPKG_LIBRARY_LINKAGE static)
+
 vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
I think this wrong.

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+# Download the Rerun SDK source archive
```
```suggestion
# Must download SDK distfile because some binaries are prebuilt from Rust.
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+# Download the Rerun SDK source archive
+vcpkg_download_distfile(
+    ARCHIVE
+    URLS "https://github.com/rerun-io/rerun/releases/download/${VERSION}/rerun_cpp_sdk.zip" # Replace with the actual URL of the zip file
```
```suggestion
    URLS "https://github.com/rerun-io/rerun/releases/download/${VERSION}/rerun_cpp_sdk.zip"
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    SHA512 1351dd0937d6ddf73622b69a803a7233eb92e5ec52607fc1c775accd015d52eaf3259c0aea64cfac3109f1c55218fb6a4597bff5b067ccdd194cd8695b3f4c8c
+)
+
+# Extract the downloaded archive
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    ARCHIVE "${ARCHIVE}"
+)
+
+# Check if the build is static or shared
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+    -DRERUN_DOWNLOAD_AND_BUILD_ARROW=OFF # Disable downloading and building Arrow
+    -DRERUN_ARROW_LINK_SHARED=${ARROW_LINK_SHARED} # Enable shared linking if not static
```
```suggestion
        -DRERUN_DOWNLOAD_AND_BUILD_ARROW=OFF # Disable downloading and building Arrow
        -DRERUN_ARROW_LINK_SHARED=${ARROW_LINK_SHARED} # Enable shared linking if not static
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    -DRERUN_ARROW_LINK_SHARED=${ARROW_LINK_SHARED} # Enable shared linking if not static
+)
+
+# Install the built files into the vcpkg installation directory
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+# Install the built files into the vcpkg installation directory
+vcpkg_cmake_install()
+
+# Fix up the CMake configuration files for compatibility with vcpkg
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    message(FATAL_ERROR "Unsupported platform")
+endif()
+
+# Replace the string for the librerun_c library in the rerun_sdkConfig.cmake file
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    "${SOURCE_PATH}/lib/${LIBRERUN_C_FILE}"
+    "\${CURRENT_PACKAGES_DIR}/lib/${LIBRERUN_C_FILE}"
+)
+
+# Replace the RERUN_LIB_DIR variable in the rerun_sdkConfig.cmake file
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    "set(RERUN_LIB_DIR \"\${CMAKE_CURRENT_LIST_DIR}/../../lib\")"
+)
+
+# Install license files for the Rerun SDK
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+    "${SOURCE_PATH}/LICENSE-APACHE"
+)
+
+# Install the usage file to provide information about how to use the library
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+# Install the usage file to provide information about how to use the library
+file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+
+# Cleanup unnecessary debug files from the installation directory
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/usage:
```diff
@@ -0,0 +1,4 @@
+rerun-sdk provides CMake targets:
+
+    find_package(rerun_sdk CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE rerun_sdk)
```
```suggestion
  find_package(rerun_sdk CONFIG REQUIRED)
  target_link_libraries(main PRIVATE rerun_sdk)
```
Also, with a line break at the end of the file. (Is a blank new line in VS Code.)

dg0yt reviewed (commented):
Comments can be useful, but there is no need to state the obvious.

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+# Check if the build is static or shared
+string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" ARROW_LINK_SHARED)
+
+# Configure the build system using CMake
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+# Fix up the CMake configuration files for compatibility with vcpkg
+vcpkg_cmake_config_fixup(PACKAGE_NAME rerun_sdk CONFIG_PATH "lib/cmake/rerun_sdk")
+
+# Determine the correct library file based on the platform and architecture
```
```suggestion
```

dg0yt commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,83 @@
+vcpkg_cmake_config_fixup(PACKAGE_NAME rerun_sdk CONFIG_PATH "lib/cmake/rerun_sdk")
+
+# Determine the correct library file based on the platform and architecture
+if(VCPKG_TARGET_IS_WINDOWS)
+    if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+        set(LIBRERUN_C_FILE "rerun_c__win_x64.lib")
+    else()
+        message(FATAL_ERROR "Unsupported Windows architecture: ${VCPKG_TARGET_ARCHITECTURE}")
+    endif()
+elseif(VCPKG_TARGET_IS_LINUX)
+    if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+        set(LIBRERUN_C_FILE "librerun_c__linux_x64.a")
+    elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+        set(LIBRERUN_C_FILE "librerun_c__linux_arm64.a")
+    else()
+        message(FATAL_ERROR "Unsupported Linux architecture: ${VCPKG_TARGET_ARCHITECTURE}")
+    endif()
+elseif(VCPKG_TARGET_IS_OSX)
+    if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+        set(LIBRERUN_C_FILE "librerun_c__macos_x64.a")
+    elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+        set(LIBRERUN_C_FILE "librerun_c__macos_arm64.a")
+    else()
+        message(FATAL_ERROR "Unsupported macOS architecture: ${VCPKG_TARGET_ARCHITECTURE}")
+    endif()
+else()
+    message(FATAL_ERROR "Unsupported platform")
+endif()
```
Untested:
```suggestion
file(GLOB LIBRERUN_C_FILE
    RELATIVE "${CURRENT_PACKAGES_DIR}/lib" 
    "${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}rerunc_c_-*${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
)
```

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/usage

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update ports/rerun-sdk/portfile.cmake

Co-authored-by: Kai Pastor <dg0yt@darc.de>

adolfos94 committed:
Update portfile.cmake

adolfos94 committed:
Add format

adolfos94 committed:
Fix identation portfile.cmake

BillyONeal commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,50 @@
+    ARCHIVE "${ARCHIVE}"
+)
+
+string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" ARROW_LINK_SHARED)
```
I filed PRs upstream that fix these:

https://github.com/rerun-io/rerun/pull/9548

https://github.com/rerun-io/rerun/pull/9550

BillyONeal commented on ports/rerun-sdk/portfile.cmake:
```diff
@@ -0,0 +1,50 @@
+vcpkg_replace_string(
+    "${CURRENT_PACKAGES_DIR}/share/rerun_sdk/rerun_sdkConfig.cmake"
+    "${SOURCE_PATH}/lib/${LIBRERUN_C_FILE}"
+    "\${CURRENT_PACKAGES_DIR}/lib/${LIBRERUN_C_FILE}"
```
`${CURRENT_PACKAGES_DIR}` doesn't exist when a config is parsed later; that's strictly a 'building in vcpkg' concept. I think this needs to be `${CMAKE_CURRENT_LIST_DIR}/../../` like below.

BillyONeal committed:
Use arrow linkage as arrow was built rather than trying to rederive it.

Incorporates https://github.com/rerun-io/rerun/pull/9548 and https://github.com/rerun-io/rerun/pull/9550

After these changes I tested that it built OK with https://github.com/rerun-io/rerun/blob/main/examples/cpp/minimal/main.cpp :

```
C:\Dev\test>dir
 Volume in drive C is OS
 Volume Serial Number is 9288-8F79

 Directory of C:\Dev\test

04/08/2025  10:42 AM    <DIR>          .
04/08/2025  10:40 AM    <DIR>          ..
04/08/2025  10:38 AM               180 CMakeLists.txt
04/08/2025  10:37 AM               756 main.cpp
               2 File(s)            936 bytes
               2 Dir(s)  4,813,367,603,200 bytes free

C:\Dev\test>type CMakeLists.txt
cmake_minimum_required(VERSION 3.10)
project(HelloWorld)

add_executable(main main.cpp)

find_package(rerun_sdk CONFIG REQUIRED)
target_link_libraries(main PRIVATE rerun_sdk)

C:\Dev\test>type main.cpp
#include <rerun.hpp>
#include <rerun/demo_utils.hpp>

using rerun::demo::grid3d;

int main() {
    // Create a new `RecordingStream` which sends data over gRPC to the viewer process.
    const auto rec = rerun::RecordingStream("rerun_example_cpp");
    // Try to spawn a new viewer instance.
    rec.spawn().exit_on_failure();

    // Create some data using the `grid` utility function.
    std::vector<rerun::Position3D> points = grid3d<rerun::Position3D, float>(-10.f, 10.f, 10);
    std::vector<rerun::Color> colors = grid3d<rerun::Color, uint8_t>(0, 255, 10);

    // Log the "my_points" entity with our data, using the `Points3D` archetype.
    rec.log("my_points", rerun::Points3D(points).with_colors(colors).with_radii({0.5f}));
}

```

BillyONeal reviewed (changes requested):
I submitted a PR which fixes these comments: https://github.com/adolfos94/vcpkg/pull/1

adolfos94 committed:
Merge pull request #1 from BillyONeal/pr-44854-rerun-sdk

Use arrow linkage as arrow was built rather than trying to rederive it.

adolfos94 reviewed (commented):
Thanks for doing this
